### PR TITLE
BAU: Add VS Code Debugging for Docker Runner

### PR DESCRIPTION
**Pull request by MarcUsher**
### Change description
- Add new script for running docker to allow VS Code which will pass variables to docker-compose file
- Update docker-compose file with default variables for normal docker running
- Add instructions to README for VS Code debugging

- [X] README and other documentation has been updated / added (if needed)
- [X] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### How to test
Should run as before, only affects local development.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,3 +37,53 @@ for authenticator, and then boots the rest of the containers. On windows run fro
...
+The 'no-debugger' part relates to the flask debugger, it's useful to remove this option if you want to see the stack traces for things like jinja template errors. The *submit* and *frontend* app commands also include 'no-reload' which disables the Flask auto-reload. This is needed for the VS Code debugging to work for these apps, but can be removed along with the `python -m debugpy --listen 0.0.0.0:5678 -m` if you'd rather use the Flask debugger and reinstate the auto-reload.
+
+To then expose the debug port 5678 to allow a debugger interface to connect, each app also has a (unique) port mapping, eg.:
+
+        ports:
+                - 5683:5678
+
+The individual *data-store* and *frontend* app dockerfiles are currently used in production so shouldn't be hardcoded to point to requirements-dev.txt (which installs debugpy) instead of requirements.txt. Instead, you can rebuild these services and pass in a buildarg which will use requirements-dev.txt to build the services to include debugpy:
+
+    docker compose build --build-arg REQUIREMENTS="requirements-dev.txt"
```
Comment (gidsg) [inline README.md:61]: Could we specify this in the docker-compose file itself?

Comment (MarcUsher) [reply README.md:61]: Yes very good suggestion, docs suggest you can add buildargs to the docker-compose file, will test with the docker-compose.debug file.

**Comment by gidsg** [comment]
@MarcUsher great work on this. Only thought is that the `docker-compose.debug.yml` may bitrot over time if people forget to update it alongside the current one.

I think a couple of possible workarounds to this could be:
* Could we use the ARGS mechanism to make the debug mode optionally triggered? e.g. an optional ARG e.g. `ENABLE_DEBUGGER` which would add additional bits to the `CMD` used by the Dockerfile?
or
* Default the main dockerfile to the debug style command (but document in the README how to remove that if you want to use the Flask debugger)?

**Comment by MarcUsher** [comment]
> @MarcUsher great work on this. Only thought is that the `docker-compose.debug.yml` may bitrot over time if people forget to update it alongside the current one.
> 
> I think a couple of possible workarounds to this could be:
> 
> * Could we use the ARGS mechanism to make the debug mode optionally triggered? e.g. an optional ARG e.g. `ENABLE_DEBUGGER` which would add additional bits to the `CMD` used by the Dockerfile?
>   or
> * Default the main dockerfile to the debug style command (but document in the README how to remove that if you want to use the Flask debugger)?

Thanks @gidsg - I'll try option 1 first which feels like it could be the safest option to avoid erroring containers.

File: README.md [hunk 2]
```diff
@@ -32,8 +32,62 @@ Where
...
+## vs-code-debug-script.sh
+Shell script to allow the VS Code debugger to attach to the running containers. See section below for full instructions.
+
+# Running in debug mode (VS Code)
+## Python Apps
+The containers in the docker runner can be run with python in debug mode to allow a debugger to connect. This gives instructions for connecting VS Code.
+
+### Docker
+
+To allow the VS Code debugger to work with the code executing in the docker runner, you can run the [vs-code-debug-script.sh](./scripts/vs-code-debug-script.sh) to pass some a variable to the docker-compose file which will allow the debugger to run.
```
Comment (gidsg) [inline README.md:49]: small typo
```suggestion
To allow the VS Code debugger to work with the code executing in the docker runner, you can run the [vs-code-debug-script.sh](./scripts/vs-code-debug-script.sh) to pass a variable to the docker-compose file which will allow the debugger to run. 
```

**Pull request merged**